File: sdks/swift/ui/Sources/VisionAgentsUI/AgentStatusView.swift
import SwiftUI
import VisionAgentsCore

/// What the agent is doing, in a line.
public struct AgentStatusView: View {
    private let state: Conversation.State

    public init(state: Conversation.State) {
        self.state = state
    }

    public var body: some View {
        HStack(spacing: 6) {
            if busy {
                ProgressView().controlSize(.mini)
            }
            Text(label)
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .animation(.default, value: label)
    }

    private var busy: Bool {
        switch state {
        case .responding, .working, .listening: return true
        case .idle, .ended: return false
        }
    }

    private var label: String {
        switch state {
        case .idle: return "ready"
        case .listening: return "listening"
        case .responding: return "answering"
        case .working(let skills):
            return skills.isEmpty
                ? "thinking" : "thinking (\(skills.joined(separator: ", ")))"
        case .ended: return "the conversation ended"
        }
    }
}

File: sdks/dart/ui/lib/src/agent_status_view.dart
import 'package:flutter/material.dart';
import 'package:vision_agents_core/vision_agents_core.dart';

/// What the agent is doing, in a line.
class AgentStatusView extends StatelessWidget {
  const AgentStatusView({super.key, required this.state});

  final ConversationState state;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final busy = switch (state) {
      Responding() || Working() || Listening() => true,
      Idle() || Ended() => false,
    };
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        if (busy) ...[
          const SizedBox.square(dimension: 10, child: CircularProgressIndicator(strokeWidth: 1.5)),
          const SizedBox(width: 6),
        ],
        Text(
          label(state),
          style: theme.textTheme.bodySmall?.copyWith(color: theme.colorScheme.outline),
        ),
      ],
    );
  }

  /// The words shown for a state, for a host drawing its own status.
  static String label(ConversationState state) => switch (state) {
    Idle() => 'ready',
    Listening() => 'listening',
    Responding() => 'answering',
    Working(:final skills) when skills.isEmpty => 'thinking',
    Working(:final skills) => 'thinking (${skills.join(', ')})',
    Ended() => 'the conversation ended',
  };
}
